README
NAME
    Path::FindDev - Find a development path somewhere in an upper hierarchy.

VERSION
    version 0.5.3

DESCRIPTION
    This package is mostly a glue layer around "Path::IsDev" with a few
    directory walking tricks.

        use Path::FindDev qw( find_dev );

        if ( my $root = find_dev('/some/path/to/something/somewhere')) {
            print "development root = $root";
        } else {
            print "No development root :(";
        }

FUNCTIONS
  find_dev
        my $result = find_dev('/some/path');

    If a "dev" directory is found at, or above, "/some/path", it will be
    returned as a "Path::Tiny"

    If you pass configurations to import:

        use Path::FindDev find_dev => { set => $someset };

    Then the exported "find_dev" will pass that set name to "Path::IsDev".

    Though you should only do this if

    *   the default set is inadequate for your usage

    *   you don't want the set to be overridden by %ENV

    Additionally, you can call find_dev directly:

        require Path::FindDev;

        my $result = Path::FindDev::find_dev('/some/path');

    Which by design inhibits your capacity to specify an alternative set in
    code.

EXAMPLE USE-CASES
    Have you ever found yourself doing

        use FindBin;
        use lib "$FindBin::Bin/../../../tlib"

    In a test?

    Have you found yourself paranoid of file-system semantics and tried

        use FindBin;
        use Path::Tiny qw(path)
        use lib path($FindBin::Bin)->parent->parent->parent->child('tlib')->stringify;

    Have you ever done either of the above in a test, only to find you've
    needed to move the test to a deeper hierarchy, and thus, need to
    re-write all your path resolution?

    Have you ever had this problem for multiple files?

    No more!

        use FindBin;
        use Path::FindDev qw(find_dev);
        use lib find_dev($FindBin::Bin)->child('t','tlib')->stringify;

    ^ Should work, regardless of which test you put it in, and regardless of
    what $CWD happens to be when you call it.

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Path-FindDev

v0.5.4 2017-03-10T07:04:44Z 5bbf076

v0.5.3 2017-03-10T07:02:09Z c9c9198
 - Bugfix: Avoid test failures due to -Ddefault_inc_excludes_dot
 - Add "v" to version due to upstream requirements
 - Tests made more portable to older Test::More

 [Dependencies::Stats]
 - Dependencies changed since 0.5.2, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: ↓1 -1 (recommends: +1)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

0.5.2 2014-08-16T04:12:22Z
 [00 Trivial]
 - CPANDAY!
 - No code changes.
 - metadata updated.
 - tests updated.
 - whitespace reduced.

 [Dependencies::Noteworthy]
 - use newer Path::Tiny to avoid bug GH #7 ( Thanks Chorny )

 [Dependencies::Stats]
 - Dependencies changed since 0.5.1, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑1)
 - runtime: ↑1
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - move to kentnl org.
 - Show 2x contributors

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - updated Test::ReportPrereqs

0.5.1 2014-07-13T17:54:11Z
 [00 Trivial]
 - Misc packaging updates.

 [Dependencies::Stats]
 - Dependencies changed since 0.5.0, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - test: (recommends: ↑2)

0.5.0 2014-03-04T11:03:22Z
 [00 Tooling release]
 - Toolkit switched to EUMM
 - No other noteworthy changes
 - Still as broken as always

 [Dependencies::Aggregate]
 - 5.008 - 49 → 44
 - 5.010 - 32 → 33
 - 5.019 - 18 → 17

 [Dependencies::Stats]
 - Dependencies changed since 0.4.2, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +8 ↑1 -10
 - runtime: +1 ↑1
 - test: +4 ↓1 -3 (recommends: +4)

0.4.2 2013-11-09T16:06:27Z
 [00 Bugfix]
 - Minimum perl declared is now 5.8, and tested to work on 5.8
 - However, the version scheme is x.y.z still, which means if you want to depend on a specific version in Perl code,
   you'll need a recent enough version.pm to make it work.

 [Dependencies::Stats]
 - Dependencies changed since 0.4.1, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +42 ↑1
 - runtime: ↓1
 - test: ↑1

0.4.1 2013-10-23T19:08:22Z
 [00 Bugfix]
 - System path test added in 0.4.0 disabled due to potential security risks ( fixes #4 )

 [Dependencies::Stats]
 - Dependencies changed since 0.4.0, see misc/*.deps* for details
 - develop: ↑2
 - test: ↑1

0.4.0 2013-10-01T22:51:43Z
 [00 API Backwards Incompatible Changes]
 - os_root property no longer exists, as is no longer required internally

 [Bugfixes]
 - Minor typo in POD ( Thanks David Steinbrunner )
 - Uplevel mechanism bugged on Win32, thanks zdm for reporting
 - Traversal Order is now test/iterate instead of iterate/test, meaning it will test / prior to terminating, useful
   probably if $PROJECTDIR is a UNC Share.
 - Root check now performed by a more reliable Path::Tiny->is_rootdir, xdg++ for the fixes.
 - Path resolution is now performed and realpathed early on.

 [Dependencies::Stats]
 - Dependencies changed since 0.3.2, see misc/*.deps* for details
 - develop: +1 -1 (recommends: -1, suggests: +1 -1)
 - runtime: ↑1 -1

 [Tests]
 - Aggressively now checks that $PROJECTROOT/../ is not contained in a directory that is considered "dev" by
   Path::IsDev.
 - This is intended to smoke out quirks that exist on your filesystem, and on CPAN Testers file systems in general that
   exhibit wart-like behaviour to the Path::IsDev Heuristic sets.
 - If these tests fail for you, they'll probably result in Path::IsDev and Path::FindDev also being broken, so I
   greatly appreciate any reports regarding false-positives that could be logically anticipated to be false positives
   on all filesystems.
 - Otherwise, please read the Path::IsDev documentation and see how to mark a directory tree as a "NegativeHeuristic",
   namely, touch $PROBLEMDIR/.path_isdev_ignore

0.3.2 2013-09-26T07:21:34Z
 [Bugfix]
 - use a less broken Test::Compile for Win32 folk

 [Dependencies::Stats]
 - Dependencies changed since 0.3.1, see misc/*.deps* for details
 - test: +1

0.3.1 2013-09-18T18:28:55Z
 [Bugfix]
 - Use Class::Tiny without 'used only once'

 [Dependencies::Stats]
 - Dependencies changed since 0.3.0, see misc/*.deps* for details
 - runtime: ↑1

0.3.0 2013-09-18T03:06:21Z
 [00 Major Changes]
 - Implemented in Class::Tiny instead, which is faster to install

 [00 Minor Changes]
 - MetaPOD info in dist.

 [Dependencies::Stats]
 - Dependencies changed since 0.2.0, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +1 (suggests: ↑1)
 - runtime: +1 -1
 - test: +2 -2

0.2.0 2013-07-07T16:36:18Z
 [00 Major Changes]
 - Refactored guts of exported symbols in terms of a Moo class.

 [Dependencies::Stats]
 - Dependencies changed since 0.1.0, see misc/*.deps* for details
 - runtime: +4 ↑1

0.1.0 2013-07-06T17:00:16Z
 - First version.


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.4008"
  },
  "configure_requires" => {
    "Module::Build" => "0.4008"
  },
  "dist_abstract" => "Find a development path somewhere in an upper hierarchy.",
  "dist_author" => [
    "Kent Fredric <kentfredric\@gmail.com>"
  ],
  "dist_name" => "Path-FindDev",
  "dist_version" => "0.4.2",
  "license" => "perl",
  "module_name" => "Path::FindDev",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Class::Tiny" => "0.010",
    "Path::IsDev" => "v0.2.2",
    "Path::IsDev::Object" => 0,
    "Path::Tiny" => "0.038",
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "File::Spec" => 0,
    "FindBin" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => "1.001002"
  }
);


my %fallback_build_requires = (
  "File::Spec" => 0,
  "FindBin" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Module::Build" => "0.4008",
  "Test::More" => "1.001002"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Path/FindDev.pm
use 5.006;    # our
use strict;
use warnings;

package Path::FindDev;

our $VERSION = 'v0.5.4';

# ABSTRACT: Find a development path somewhere in an upper hierarchy.

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

use Sub::Exporter -setup => { exports => [ find_dev => \&_build_find_dev, ] };












sub _build_find_dev {
  my ( undef, undef, $arg ) = @_;

  my $finddev_object;
  return sub {
    my ($path) = @_;
    $finddev_object ||= do {
      require Path::FindDev::Object;
      Path::FindDev::Object->new($arg);
    };
    return $finddev_object->find_dev($path);
  };
}


































*find_dev = _build_find_dev( __PACKAGE__, 'find_dev', {} );

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Path::FindDev - Find a development path somewhere in an upper hierarchy.

=head1 VERSION

version v0.5.4

=head1 DESCRIPTION

This package is mostly a glue layer around L<< C<Path::IsDev>|Path::IsDev >>
with a few directory walking tricks.

    use Path::FindDev qw( find_dev );

    if ( my $root = find_dev('/some/path/to/something/somewhere')) {
        print "development root = $root";
    } else {
        print "No development root :(";
    }

=head1 FUNCTIONS

=head2 find_dev

    my $result = find_dev('/some/path');

If a C<dev> directory is found at, or above, C</some/path>, it will be returned
as a L<< C<Path::Tiny>|Path::Tiny >>

If you pass configurations to import:

    use Path::FindDev find_dev => { set => $someset };

Then the exported C<find_dev> will pass that set name to L<< C<Path::IsDev>|Path::IsDev >>.

Though you should only do this if

=over 4

=item * the default set is inadequate for your usage

=item * you don't want the set to be overridden by C<%ENV>

=back

Additionally, you can call find_dev directly:

    require Path::FindDev;

    my $result = Path::FindDev::find_dev('/some/path');

Which by design inhibits your capacity to specify an alternative set in code.

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Path::FindDev",
    "interface":"exporter"
}


=end MetaPOD::JSON

=head1 EXAMPLE USE-CASES

Have you ever found yourself doing

    use FindBin;
    use lib "$FindBin::Bin/../../../tlib"

In a test?

Have you found yourself paranoid of file-system semantics and tried

    use FindBin;
    use Path::Tiny qw(path)
    use lib path($FindBin::Bin)->parent->parent->parent->child('tlib')->stringify;

Have you ever done either of the above in a test, only to
find you've needed to move the test to a deeper hierarchy,
and thus, need to re-write all your path resolution?

Have you ever had this problem for multiple files?

No more!

    use FindBin;
    use Path::FindDev qw(find_dev);
    use lib find_dev($FindBin::Bin)->child('t','tlib')->stringify;

^ Should work, regardless of which test you put it in, and regardless
of what C<$CWD> happens to be when you call it.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint-travis-ci/script.pl
#!/usr/bin/env perl

use strict;
use warnings;
use utf8;

sub diag { print STDERR @_; print STDERR "\n" }
sub env_exists { return exists $ENV{ $_[0] } }
sub env_true { return env_exists( $_[0] ) and $ENV{ $_[0] } }
sub env_is { return env_exists($_[0]) and $ENV{$_[0]} eq $_[1] }

sub safe_exec {
  my ( $command, @params ) = @_;
  diag("running $command @params");
  my $exit = system( $command, @params );
  if ( $exit != 0 ) {
    my $low  = $exit & 0b11111111;
    my $high = $exit >> 8;
    warn "$command failed: $? $! and exit = $high , flags = $low";
    if ( $high != 0 ) {
      exit $high;
    }
    else {
      exit 1;
    }
  }
  return 1;
}

if ( not env_exists('TRAVIS') ) {
  diag('Is not running under travis!');
  exit 1;
}

if ( env_is('TRAVIS_BRANCH', 'master' ) ) {

} else {
    my @paths = './t';

    if (  env_true('AUTHOR_TESTING') or env_true('RELEASE_TESTING') ) {
      push @paths, './xt';
    }
    safe_exec( 'prove', '--blib', '--shuffle', '--color', '--recurse', '--timer', '--jobs', 30, @paths );
}


lib/Path/FindDev/Object.pm
use 5.006;    # our
use strict;
use warnings;

package Path::FindDev::Object;

our $VERSION = 'v0.5.4';

# ABSTRACT: Object oriented guts to FindDev

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY

our $ENV_KEY_DEBUG = 'PATH_FINDDEV_DEBUG';
our $DEBUG = ( exists $ENV{$ENV_KEY_DEBUG} ? $ENV{$ENV_KEY_DEBUG} : undef );













use Class::Tiny 0.010 'set', 'uplevel_max', {
  nest_retry => sub {
    return 0;
  },
  isdev => sub {
    require Path::IsDev::Object;
    return Path::IsDev::Object->new( ( $_[0]->has_set ? ( set => $_[0]->set ) : () ) );
  },
};









## no critic (RequireArgUnpacking)







sub has_set { return exists $_[0]->{set} }















sub has_uplevel_max { return exists $_[0]->{uplevel_max} }


















my $instances   = {};
my $instance_id = 0;














sub _instance_id {
  my ($self) = @_;
  require Scalar::Util;
  my $addr = Scalar::Util::refaddr($self);
  return $instances->{$addr} if exists $instances->{$addr};
  $instances->{$addr} = sprintf '%x', $instance_id++;
  return $instances->{$addr};
}










sub BUILD {
  my ($self) = @_;
  return $self unless $DEBUG;
  $self->_debug('{');
  $self->_debug( '  set         => ' . $self->set )         if $self->has_set;
  $self->_debug( '  uplevel_max => ' . $self->uplevel_max ) if $self->uplevel_max;
  $self->_debug( '  nest_retry  => ' . $self->nest_retry );
  $self->_debug( '  isdev       => ' . $self->isdev );
  $self->_debug('}');
  return $self;
}













sub _debug {
  my ( $self, $message ) = @_;
  return unless $DEBUG;
  my $id = $self->_instance_id;
  return *STDERR->printf( qq{[Path::FindDev=%s] %s\n}, $id, $message );
}









sub _error {
  my ( $self, $message ) = @_;
  my $id = $self->_instance_id;
  my $f_message = sprintf qq{[Path::FindDev=%s] %s\n}, $id, $message;
  require Carp;
  Carp::croak($f_message);
}
















sub _step {
  my ( $self, $search_root, $dev_levels, $uplevels ) = @_;

  if ( $self->isdev->matches($search_root) ) {
    $self->_debug( 'Found dev dir' . $search_root );
    ${$dev_levels}++;
    return { type => 'found', path => $search_root } if ${$dev_levels} >= $self->nest_retry;
    $self->_debug( sprintf 'Ignoring found dev dir due to dev_levels(%s) < nest_retry(%s)', ${$dev_levels}, $self->nest_retry );
  }
  if ( $search_root->is_rootdir ) {
    $self->_debug('OS Root hit ( ->is_rootdir )');
    return { type => 'stop' };
  }
  if ( $self->has_uplevel_max and ${$uplevels} > $self->uplevel_max ) {
    $self->_debug( 'Stopping search due to uplevels(%s) >= uplevel_max(%s)', ${$uplevels}, $self->uplevel_max );
    return { type => 'stop' };
  }

  return { type => 'next' };
}









sub find_dev {
  my ( $self, $path ) = @_;
  require Path::Tiny;
  my $search_root = Path::Tiny::path($path)->absolute->realpath;
  $self->_debug( 'Finding dev for ' . $path );
  my $dev_levels = 0;
  my $uplevels   = 0 - 1;
FLOW: {
    $uplevels++;
    my $result = $self->_step( $search_root, \$dev_levels, \$uplevels );
    if ( 'next' eq $result->{type} ) {
      $self->_debug( 'Trying ../ : ' . $search_root->parent );
      $search_root = $search_root->parent;
      redo FLOW;
    }
    if ( 'stop' eq $result->{type} ) {
      return;
    }
    if ( 'found' eq $result->{type} ) {
      return $result->{path};
    }
    $self->_error( 'Unexpected end of flow control with _step response type' . $result->{type} );
  }
  return;
}
1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Path::FindDev::Object - Object oriented guts to FindDev

=head1 VERSION

version v0.5.4

=head1 SYNOPSIS

    require Path::FindDev::Object;
    my $finder = Path::FindDev::Object->new();
    my $dev = $finder->find_dev($path);

=head1 DESCRIPTION

This module implements the innards of L<< C<Path::FindDev>|Path::FindDev >>, and is
only recommended for use if the Exporter C<API> is insufficient for your needs.

=head1 METHODS

=head2 C<has_set>

Determines if the C<set> attribute exists

=head2 C<has_uplevel_max>

Determines if the C<uplevel_max> attribute is provided.

=head2 C<find_dev>

Find a parent at, or above C<$OtherPath> that resembles a C<devel> directory.

    my $path = $object->find_dev( $OtherPath );

=head1 ATTRIBUTES

=head2 C<set>

B<(optional)>

The C<Path::IsDev::HeuristicSet> subclass for your desired Heuristics.

=head2 C<uplevel_max>

If provided, limits the number of C<uplevel> iterations done.

( that is, limits the number of times it will step up the hierarchy )

=head2 C<nest_retry>

The number of C<dev> directories to C<ignore> in the hierarchy.

This is provided in the event you have a C<dev> directory within a C<dev> directory, and you wish
to resolve an outer directory instead of an inner one.

By default, this is C<0>, or "stop at the first C<dev> directory"

=head2 C<isdev>

The L<< C<Path::IsDev>|Path::IsDev >> object that checks nodes for C<dev>-ishness.

=head1 PRIVATE METHODS

=head2 C<_instance_id>

An opportunistic sequence number for help with debug messages.

Note: This is not guaranteed to be unique per instance, only guaranteed
to be constant within the life of the object.

Based on C<refaddr>, and giving out new ids when new C<refaddr>'s are seen.

    my $id = $object->_instance_id;

=head2 C<BUILD>

C<BUILD> is an implementation detail of C<Moo>/C<Moose>.

This module hooks C<BUILD> to give a self report of the object
to C<*STDERR> after C<< ->new >> when under C<$DEBUG>

=head2 C<_debug>

The debugger callback.

    export PATH_FINDDEV_DEBUG=1

to get debug info.

    $object->_debug($message);

=head2 C<_error>

The error reporting callback.

    $object->_error($message);

=head2 C<_step>

Inner code path of tree walking.

    my ($dev_levels, $uplevels ) = (0,0);

    my $result = $object->_step( path($somepath), \$dev_levels, \$uplevels );

    $result->{type} eq 'stop'   # if flow control should end
    $result->{type} eq 'next'   # if flow control should ascend to parent
    $result->{type} eq 'found'  # if flow control has found the "final" dev directory

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Path::FindDev::Object",
    "interface":"class",
    "inherits":"Class::Tiny::Object"
}


=end MetaPOD::JSON

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


maint/perlcritic.rc.gen.pl
#!/usr/bin/env perl
# FILENAME: bundle_to_ini.pl
# CREATED: 02/06/14 01:48:56 by Kent Fredric (kentnl) <kentfredric@gmail.com>
# ABSTRACT: Write an INI file from a bundle

use strict;
use warnings;
use utf8;

use Carp qw( croak carp );
use Perl::Critic::ProfileCompiler::Util qw( create_bundle );

my $bundle = create_bundle('Example::Author::KENTNL');
$bundle->configure;

my @stopwords = (qw( FindDev ));
for my $var ( sort @stopwords ) {
  $bundle->add_or_append_policy_field( 'Documentation::PodSpelling' => ( 'stop_words' => $var ) );
}

#$bundle->add_or_append_policy_field(
#  'Subroutines::ProhibitCallsToUndeclaredSubs' => ( 'exempt_subs' => 'String::Formatter::str_rf' ), );
#
#$bundle->remove_policy('ErrorHandling::RequireCarping');
#$bundle->remove_policy('NamingConventions::Capitalization');
$bundle->remove_policy('ValuesAndExpressions::RequireNumericVersion');

my $inf = $bundle->actionlist->get_inflated;

my $config = $inf->apply_config;
{
  open my $rcfile, '>', './perlcritic.rc' or croak 'Cant open perlcritic.rc';
  $rcfile->print( $config->as_ini, "\n" );
  close $rcfile or croak 'Something fubared closing perlcritic.rc';
}
my $deps = $inf->own_deps;
{
  open my $depsfile, '>', './perlcritic.deps' or croak 'Cant open perlcritic.deps';
  for my $key ( sort keys %{$deps} ) {
    $depsfile->printf( "%s~%s\n", $key, $deps->{$key} );
    *STDERR->printf( "%s => %s\n", $key, $deps->{$key} );
  }
  close $depsfile or carp 'Something fubared closing perlcritic.deps';
}



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Find a development path somewhere in an upper hierarchy.",
  "AUTHOR" => "Kent Fredric <kentnl\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Path-FindDev",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Path::FindDev",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Tiny" => "0.010",
    "Path::IsDev" => "0.2.2",
    "Path::IsDev::Object" => 0,
    "Path::Tiny" => "0.054",
    "Scalar::Util" => 0,
    "Sub::Exporter" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "FindBin" => 0,
    "Test::More" => 0
  },
  "VERSION" => "v0.5.4",
  "test" => {
    "TESTS" => "t/*.t t/00-compile/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Tiny" => "0.010",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "FindBin" => 0,
  "Path::IsDev" => "0.2.2",
  "Path::IsDev::Object" => 0,
  "Path::Tiny" => "0.054",
  "Scalar::Util" => 0,
  "Sub::Exporter" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
